doc/developer/design/20230216_role_based_access_control.md

```diff
@@ -0,0 +1,440 @@
+    - `TRIGGER`
+    - `CONNECT`
+    - `TEMPORARY`
+    - `EXECUTE`
+    - `SET`
+    - `ALTER SYSTEM`
+- The following object types:
+    - `DOMAIN`
+    - `FUNCTION` or `PROCEDURE`
+    - `FOREIGN DATA WRAPPER`
+    - `FOREIGN SERVER`
+    - `LANGUAGE`
+    - `LARGE OBJECT`
+    - `PARAMETER`
+    - `SEQUENCE`
+    - `Table column`
+    - `TABLESPACE`
+    - `TYPE`
+- Adding the necessary pg views to support all role based `psql` meta-commands.
+
+### Phase 5 - Utility Commands (Optional)
+
+This is an optional phase to add some utility commands present in PostgreSQL. We will add the following SQL commands:
+
+- `REASSIGN OWNED BY <old_role> TO <new_role>`
+    - Transfers ownership of all objects owned by `<old_role>` to `<new_role>`.
+    - Can only be run by a member of `<old_role>` and `<new_role>` or a superuser.
+    - Requires `CREATE` privilege on all schemas and databases where all objects reside.
+        - TODO: This isn't explicitly stated in th docs, but would make sense base on the `ALTER` privileges. I will
+          double-check this.
+    - In PostgreSQL, this only affects the current database, and does not reassign the database itself. We will diverge
+      here and have it affect all databases, including the databases themselves.
+- `DROP OWNED BY <name> [ CASCADE | RESTRICT]`
+    - Drops all objects owned by `<name>`.
+    - Requires membership of `<name>`.
+    - In PostgreSQL, this only affects the current database, and does not drop the database itself. We will diverge here
+      and have it affect all databases, including the databases themselves.
+    - Revokes all privileges granted to `<name>`.
+    - Default is `RESTRICT`.
+
+#### Out of Scope for Project
+
+- `CURRENT_ROLE`, `CURRENT_USER`, `SESSION_USER`, aliases in `GRANT`, `REVOKE`, `ALTER`, `REASSIGN OWNED`
+  and `DROP OWNED`.
+
+## Rollout Plan
+
+- All new SQL commands will be gated by unsafe mode (or LaunchDarkly depending on the standard for feature flags when
+  this is implemented).
+- Specifying attributes in `CREATE ROLE` will be gated by unsafe mode (or LaunchDarkly depending on the standard for
+  feature flags when this is implemented).
+- All roles that authenticate through Frontegg will be given the `SUPERUSER` session attribute. This matches the
+  existing behavior where all roles have the `SUPERUSER` attribute.
+- Roles that are not authenticated through Frontegg, will not be given the `SUPERUSER` session attribute. This allows us
+  to test various permissions locally.
+- When we are ready to release this feature, we will do the following:
+    - Remove the unsafe gate from new SQL commands and role attributes.
+    - Modify the Frontegg authentication so that only roles that are Frontegg admins are given the `SUPERUSER` session
+      attribute.
+
+## Testing Plan
+
+- A lot of SLT and Testdrive tests that create various roles, grant them various attributes and privileges, and test
+  what they are and are not allowed to do.
```

Comment:
Unless sqllogictest already supports the required functionality, we may need to extend testdrive in order to do the following:
- connecting using some username/password
- issue a query and expect an error

Currently in testdrive, you can either issue queries with the default username and issue queries that are expected to error out *or* connect with a different username and only issue queries that are expected to succeed.

Reply:
sqllogictest can do this! https://github.com/MaterializeInc/materialize/blob/main/doc/developer/sqllogictest.md#simple-extension See the mz_system exception.